File: frontend/lib/services/weather_service.dart
// Copyright (c) 2025 CuongKenn and ICTU-OpenAgri Contributors
// Licensed under the MIT License. See LICENSE file in the project root for full license information.

import 'package:dio/dio.dart';
import 'package:geolocator/geolocator.dart';
import '../models/api_models.dart';
import 'api_service.dart';
import 'auth_service.dart';

class WeatherService {
  // Singleton pattern
  static final WeatherService _instance = WeatherService._internal();
  factory WeatherService() => _instance;
  WeatherService._internal();

  final ApiService _apiService = ApiService();
  final AuthService _authService = AuthService();

  // Get current location
  Future<Position> getCurrentLocation() async {
    bool serviceEnabled;
    LocationPermission permission;

    serviceEnabled = await Geolocator.isLocationServiceEnabled();
    if (!serviceEnabled) {
      throw Exception('Location services are disabled.');
    }

    permission = await Geolocator.checkPermission();
    if (permission == LocationPermission.denied) {
      permission = await Geolocator.requestPermission();
      if (permission == LocationPermission.denied) {
        throw Exception('Location permissions are denied');
      }
    }

    if (permission == LocationPermission.deniedForever) {
      throw Exception(
          'Location permissions are permanently denied, we cannot request permissions.');
    }

    // Try to get last known position first for speed
    try {
      final lastKnown = await Geolocator.getLastKnownPosition();
      if (lastKnown != null) {
        return lastKnown;
      }
    } catch (e) {
      // Ignore error and try current position
    }

    return await Geolocator.getCurrentPosition();
  }

  // Search location using Backend
  Future<List<Map<String, dynamic>>> searchLocation(String query) async {
    if (query.length < 3) return [];

    try {
      final token = await _authService.getToken();
      final response = await _apiService.client.get(
        '/weather/search',
        queryParameters: {
          'query': query,
          'limit': 5,
        },
        options: Options(
          headers: {'Authorization': 'Bearer $token'},
        ),
      );

      final dto = LocationSearchResponseDTO.fromJson(response.data);

      return dto.results.map((item) {
        return {
          'name': item.name,
          'city': item.state ?? item.country,
          'country': item.country,
          'lat': item.latitude,
          'lon': item.longitude,
        };
      }).toList();
    } catch (e) {
      return [];
    }
  }

  // Get weather data from Backend
  Future<Map<String, dynamic>> getWeatherData(double lat, double lon) async {
    try {
      final token = await _authService.getToken();
      final response = await _apiService.client.get(
        '/weather/forecast',
        queryParameters: {
          'latitude': lat,
          'longitude': lon,
          'hours_ahead': 168,
        },
        options: Options(
          headers: {'Authorization': 'Bearer $token'},
        ),
      );

      final dto = ForecastResponseDTO.fromJson(response.data);
      return _transformBackendData(dto);
    } catch (e) {
      // Return mock data if API fails
      return _getMockWeatherData();
    }
  }

  Map<String, dynamic> _transformBackendData(ForecastResponseDTO dto) {
    final current = {
      'temperature_2m': dto.current.temperature2m,
      'relative_humidity_2m': dto.current.relativeHumidity2m,
      'weather_code': dto.current.weatherCode,
      'wind_speed_10m': dto.current.windSpeed10m,
      'precipitation': dto.current.precipitation,
      'is_day': dto.current.isDay,
    };

    final hourlyList = dto.hourly;

    // Transform hourly list to column-oriented map for UI
    final hourly = {
      'time': hourlyList.map((e) => e.time).toList(),
      'temperature_2m': hourlyList.map((e) => e.temperature2m).toList(),
      'relative_humidity_2m':
          hourlyList.map((e) => e.relativeHumidity2m).toList(),
      'weather_code': hourlyList.map((e) => e.weatherCode).toList(),
      'wind_speed_10m': hourlyList.map((e) => e.windSpeed10m).toList(),
      'precipitation_probability':
          hourlyList.map((e) => (e.precipitation > 0 ? 60 : 0)).toList(),
      'soil_moisture': hourlyList.map((e) => e.soilMoisture).toList(),
    };

    // Generate daily data from hourly (simplified aggregation)
    final daily = _generateDailyFromHourly(hourlyList);

    return {
      'current': current,
      'hourly': hourly,
      'daily': daily,
    };
  }

  Map<String, dynamic> _generateDailyFromHourly(
      List<HourlyWeatherDTO> hourlyList) {
    final List<String> times = [];
    final List<double> maxTemps = [];
    final List<double> minTemps = [];
    final List<int> codes = [];
    final List<int> probs = [];

    // Group by day
    final Map<String, List<HourlyWeatherDTO>> days = {};
    for (var item in hourlyList) {
      final timeStr = item.time;
      final dateStr = timeStr.split('T')[0]; // Assumes ISO format
      if (!days.containsKey(dateStr)) {
        days[dateStr] = [];
      }
      days[dateStr]!.add(item);
    }

    // Process each day
    days.forEach((date, items) {
      if (times.length >= 7) return;

      times.add(date);

      // Calculate max/min temp
      double maxT = -100;
      double minT = 100;
      int maxProb = 0;
      final Map<int, int> codeCounts = {};

      for (var item in items) {
        final t = item.temperature2m;
        if (t > maxT) maxT = t;
        if (t < minT) minT = t;

        final p = item.precipitation;
        if (p > 0) maxProb = 60; // Simplified

        final c = item.weatherCode;
        codeCounts[c] = (codeCounts[c] ?? 0) + 1;
      }

      maxTemps.add(maxT);
      minTemps.add(minT);
      probs.add(maxProb);

      // Most frequent weather code
      int mostFrequentCode = 0;
      int maxCount = 0;
      codeCounts.forEach((code, count) {
        if (count > maxCount) {
          maxCount = count;
          mostFrequentCode = code;
        }
      });
      codes.add(mostFrequentCode);
    });

    return {
      'time': times,
      'temperature_2m_max': maxTemps,
      'temperature_2m_min': minTemps,
      'weather_code': codes,
      'precipitation_probability_max': probs,
    };
  }

  // Helper to map WMO weather codes to description and icon
  Map<String, dynamic> getWeatherInfo(int code) {
    // WMO Weather interpretation codes (WW)
    // https://open-meteo.com/en/docs
    switch (code) {
      case 0:
        return {'desc': 'Trời quang', 'icon': 'clear_day'};
      case 1:
        return {'desc': 'Chủ yếu là nắng', 'icon': 'partly_cloudy_day'};
      case 2:
        return {'desc': 'Có mây', 'icon': 'partly_cloudy_day'};
      case 3:
        return {'desc': 'Nhiều mây', 'icon': 'cloudy'};
      case 45:
      case 48:
        return {'desc': 'Sương mù', 'icon': 'foggy'};
      case 51:
      case 53:
      case 55:
        return {'desc': 'Mưa phùn', 'icon': 'rainy_light'};
      case 61:
      case 63:
      case 65:
        return {'desc': 'Mưa', 'icon': 'rainy'};
      case 80:
      case 81:
      case 82:
        return {'desc': 'Mưa rào', 'icon': 'rainy_heavy'};
      case 95:
      case 96:
      case 99:
        return {'desc': 'Dông', 'icon': 'thunderstorm'};
      default:
        return {'desc': 'Không xác định', 'icon': 'cloud'};
    }
  }

  Map<String, dynamic> _getMockWeatherData() {
    final now = DateTime.now();
    final hourlyTimes =
        List.generate(24, (i) => now.add(Duration(hours: i)).toIso8601String());
    final dailyTimes =
        List.generate(7, (i) => now.add(Duration(days: i)).toIso8601String());

    return {
      "current": {
        "temperature_2m": 32.0,
        "relative_humidity_2m": 75,
        "apparent_temperature": 35.0,
        "is_day": 1,
        "precipitation": 0.0,
        "rain": 0.0,
        "weather_code": 1,
        "wind_speed_10m": 15.0
      },
      "hourly": {
        "time": hourlyTimes,
        "temperature_2m": List.filled(24, 30.0),
        "relative_humidity_2m": List.filled(24, 70),
        "weather_code": List.filled(24, 1),
        "wind_speed_10m": List.filled(24, 10.0),
        "precipitation_probability": List.filled(24, 0),
      },
      "daily": {
        "time": dailyTimes,
        "temperature_2m_max": List.filled(7, 35.0),
        "temperature_2m_min": List.filled(7, 25.0),
        "weather_code": List.filled(7, 1),
        "precipitation_probability_max": List.filled(7, 20),
      }
    };
  }
}

File: backend/app/infrastructure/external_services/weather_service.py
"""
Weather service for Open-Meteo and Photon API integration.
"""
import logging
from typing import Optional, List, Dict, Any

try:
    import httpx
except ImportError:
    import requests as httpx

logger = logging.getLogger(__name__)


class OpenMeteoService:
    """Service for Open-Meteo weather API."""
    
    BASE_URL = "https://api.open-meteo.com/v1"
    
    def __init__(self, timeout: int = 10):
        self.timeout = timeout
    
    async def get_forecast(
        self, 
        latitude: float, 
        longitude: float,
        hours_ahead: int = 24
    ) -> Dict[str, Any]:
        """
        Get weather forecast from Open-Meteo API.
        
        Args:
            latitude: Location latitude
            longitude: Location longitude
            hours_ahead: Number of hours to forecast (max 240)
        
        Returns:
            Weather forecast data
        """
        if hours_ahead > 240:
            hours_ahead = 240
        
        async with httpx.AsyncClient() as client:
            try:
                response = await client.get(
                    f"{self.BASE_URL}/forecast",
                    params={
                        "latitude": latitude,
                        "longitude": longitude,
                        "current": "temperature_2m,relative_humidity_2m,weather_code,wind_speed_10m,precipitation,is_day",
                        "hourly": "temperature_2m,relative_humidity_2m,weather_code,wind_speed_10m,precipitation,soil_moisture_0_to_1cm",
                        "forecast_days": (hours_ahead // 24) + 1,
                        "timezone": "auto"
                    },
                    timeout=self.timeout
                )
                response.raise_for_status()
                return response.json()
            except httpx.HTTPError as e:
                logger.error(f"Error fetching forecast from Open-Meteo: {str(e)}")
                raise Exception(f"Failed to fetch weather data: {str(e)}")
    
    async def get_historical_data(
        self,
        latitude: float,
        longitude: float,
        start_date: str,
        end_date: str
    ) -> Dict[str, Any]:
        """
        Get historical weather data.
        
        Args:
            latitude: Location latitude
            longitude: Location longitude
            start_date: Start date (YYYY-MM-DD)
            end_date: End date (YYYY-MM-DD)
        
        Returns:
            Historical weather data
        """
        async with httpx.AsyncClient() as client:
            try:
                response = await client.get(
                    f"{self.BASE_URL}/archive",
                    params={
                        "latitude": latitude,
                        "longitude": longitude,
                        "start_date": start_date,
                        "end_date": end_date,
                        "hourly": "temperature_2m,relative_humidity_2m,precipitation,wind_speed_10m",
                        "timezone": "auto"
                    },
                    timeout=self.timeout
                )
                response.raise_for_status()
                return response.json()
            except httpx.HTTPError as e:
                logger.error(f"Error fetching historical data: {str(e)}")
                raise Exception(f"Failed to fetch historical data: {str(e)}")


class PhotonGeocodingService:
    """Service for Photon API (reverse geocoding and location search)."""
    
    BASE_URL = "https://photon.komoot.io"
    
    def __init__(self, timeout: int = 10):
        self.timeout = timeout
    
    async def search_location(
        self,
        query: str,
        limit: int = 10,
        country_codes: Optional[List[str]] = None
    ) -> Dict[str, Any]:
        """
        Search for locations by query string.
        
        Args:
            query: Search query (city name, address, etc.)
            limit: Maximum number of results
            country_codes: Filter by country codes (e.g., ['vn'])
        
        Returns:
            Search results with coordinates
        """
        params = {
            "q": query,
            "limit": min(limit, 50)
        }
        
        if country_codes:
            params["osm_tag"] = ",".join(country_codes)
        
        async with httpx.AsyncClient() as client:
            try:
                response = await client.get(
                    f"{self.BASE_URL}/api",
                    params=params,
                    timeout=self.timeout
                )
                response.raise_for_status()
                return response.json()
            except httpx.HTTPError as e:
                logger.error(f"Error searching location: {str(e)}")
                raise Exception(f"Failed to search location: {str(e)}")
    
    async def reverse_geocode(
        self,
        latitude: float,
        longitude: float,
        limit: int = 1
    ) -> Dict[str, Any]:
        """
        Get location name and address from coordinates (reverse geocoding).
        
        Args:
            latitude: Location latitude
            longitude: Location longitude
            limit: Maximum number of results
        
        Returns:
            Location information
        """
        params = {
            "lat": latitude,
            "lon": longitude,
            "limit": min(limit, 10)
        }
        
        async with httpx.AsyncClient() as client:
            try:
                response = await client.get(
                    f"{self.BASE_URL}/reverse",
                    params=params,
                    timeout=self.timeout
                )
                response.raise_for_status()
                return response.json()
            except httpx.HTTPError as e:
                logger.error(f"Error reverse geocoding: {str(e)}")
                raise Exception(f"Failed to reverse geocode: {str(e)}")
